pkg/pgmodel/ingestor/metric_batcher.go

```diff
@@ -208,6 +213,11 @@ func runMetricBatcher(conn pgxconn.PgxConn,
 		}
 
 		numSeries := pending.batch.CountSeries()
+
+		// Since seriesId order is not guaranteed we need to sort it to avoid row deadlock when duplicates are sent (eg. Prometheus retry)
+		// Samples inside series should be sorted by Prometheus
+		sort.Sort(&pending.batch)
```

Comment:
We can't sort here since seriesID is not guaranteed to be set yet (it will be if there is a cache hit but not a cache miss). It's only guaranteed to be set after the WriteSeries call in persistBatch (series.go). I think we should sort there.